### ISSUE TITLE
About the configuration in sam3_concept_bank.py

### ISSUE THREAD

**Issue by SmallLeishen** [ConceptBank/ConceptBank#2]
I want to debug the whole code for generating the concept bank. But there are still some parameters whose values are not clear how to specify.
parser.add_argument("--users", type=str, help="name=cfg.py,...",  # **file1: the configuration file of the dataset in "/ConceptBank/configs"**
                        default="/ConceptBank/configs/cfg_coco_stuff10k.py")
parser.add_argument("--img_path", type=str, # **file2: the folder where the images are stored**
                        default="/ConceptBank/data/cocostuff10k/images")
parser.add_argument("--seg_path", type=str, # **file3: the folder where the annotated .mat files are stored**
                        default="/ConceptBank/data/cocostuff10k/annotations")
parser.add_argument("--checkpoint_path", type=str, # **file4: ???**
                        default="/ConceptBank/data/stare/cb_sam3_ns.pt")
parser.add_argument("--bpe_path", type=str, # **file5: ???** 
                        default="/ConceptBank/data/cocostuff10k/imageLists/all.txt")
parser.add_argument("--output_pt", type=str, # **file6: output file path**
                        default="/ConceptBank/output")
**What exactly do file4 and file5 refer to? Is the path to file1 and its content correct?**
################
file1:
_base_ = './base_config.py'

# model settings
model = dict(
    name_path='/jsj_zff/tjut_kaisenlei/ConceptBank/configs/ori_cls/cls_coco_stuff.txt',
    concept_path='/jsj_zff/tjut_kaisenlei/ConceptBank/data/stare/cb_sam3_ns.pt',
    concept_user='stuff10k',
    confidence_threshold=0.3,
)

# dataset settings
dataset_type = 'STUFF10KDataset'
data_root = '/jsj_zff/tjut_kaisenlei/ConceptBank/data/cocostuff10k'

test_pipeline = [
    dict(type='LoadPILFromFile'),
    dict(type='LoadAnnotations', reduce_zero_label=True),
    dict(type='PackPILInputs', to_tensor=False)
]

test_dataloader = dict(
    batch_size=1,
    num_workers=4,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        data_prefix=dict(
            img_path='images',
            seg_map_path='annotations'),
        pipeline=test_pipeline))
##################

**Comment by ConceptBank** [comment]
Hi, thank you for your interest in our work and for reaching out with your questions!

To reproduce our results, please strictly follow the pipeline outlined in our README. Here are the specific steps:

**1. Environment & Dataset Preparation**
First, ensure your environment is fully set up and the datasets are prepared correctly. Please follow the instructions in our [Dataset Preparation Guide](https://github.com/ConceptBank/ConceptBank?tab=readme-ov-file#iv-dataset-preparation).

**2. Concept Bank Generation**
We have provided ready-to-use script files to build the Concept Bank. You can reference the details [here](https://github.com/ConceptBank/ConceptBank?tab=readme-ov-file#concept-bank-generation-optional).

**3. Script Adjustments**
To run the generation process, open the provided `.sh` script and make the following minor adjustments:

* Uncomment the block of code for generating the Concept Bank.
* Replace `sam3_concept_bank_time.py` with `sam3_concept_bank.py`.
* All necessary parameter configurations are already included in the command.

**4. Configuration Files**
You **do not** need to manually build or modify any new configuration files. Everything you need is already provided in the `configs/` directory.

For your convenience, here is how the modified script should look:

```bash
#!/bin/bash

NGPUS=${1:-4}
LOGFILE=${2:-"logs_ns.txt"}

# Concept Bank Generation
CUDA_VISIBLE_DEVICES=0,1,2,3 \
OMP_NUM_THREADS=16 \
MKL_NUM_THREADS=1 \
TORCH_CUDNN_SDPA_ENABLED=1 \
torchrun --nproc_per_node="${NGPUS}" --master_port=29500 \
   sam3_concept_bank.py \
   --users "voc21=configs/cfg_voc21.py,
            context60=configs/cfg_context60.py,
            coco_object=configs/cfg_coco_object.py,
            voc20=configs/cfg_voc20.py,
            context59=configs/cfg_context59.py,
            coco_stuff164k=configs/cfg_coco_stuff164k.py,
            city_scapes=configs/cfg_city_scapes.py,
            ade20k=configs/cfg_ade20k.py" \
   --split train \
   --checkpoint_path sam3/assets/sam3.pt \
   --bpe_path sam3/assets/bpe_simple_vocab_16e6.txt.gz \
   --cand_topk 999 --fuse_topk 999 \
   --confidence_threshold 0.1 \
   --use_context_view --use_masked_view \
   --output_pt "./configs/concept_bank/cb_sam3_ns.pt"

# Evaluation
DATASETS=(
  voc21
  context60
  coco_object
  voc20
  context59
  coco_stuff164k
  city_scapes
  ade20k
)

: > "${LOGFILE}"

for BENCHMARK in "${DATASETS[@]}"; do
  {
    printf "\n=== DATASET: %s ===\n\n" \
      "${BENCHMARK}"
  } >> "${LOGFILE}"

  CUDA_VISIBLE_DEVICES=0,1,2,3 \
  OMP_NUM_THREADS=16 \
  MKL_NUM_THREADS=1 \
  TORCH_CUDNN_SDPA_ENABLED=1 \
  torchrun --nproc_per_node="${NGPUS}" --master_port=29500 \
    eval.py \
    --launcher pytorch \
    --config "./configs/cfg_${BENCHMARK}.py" \
    |& tee -a "${LOGFILE}"

  echo "----------" >> "${LOGFILE}"
done

python3 summarize_seg_metrics.py "${LOGFILE}"

```

**Comment by SmallLeishen** [comment]
If I want to evaluate the performance of SAM3 and SAM3 + prompt expansion, how should I set the script hyperparameters respectively?

**Comment by ConceptBank** [comment]
Hi! Thanks for your question.

Since SAM3 doesn't officially provide an Open-Vocabulary Semantic Segmentation (OVSS) inference script, the custom `SAM3_OVSS` code provided below is a great starting point.

Here is:
cfg_voc21.py

```python
model = dict(
    classname_path='./configs/cls_voc21.txt'
)

```

sam3_ovss.py
```python
import torch
from torch import nn
from typing import List, Tuple, Optional, Any
from PIL import Image

from mmengine.structures import PixelData
from mmseg.registry import MODELS
from mmseg.models.segmentors import BaseSegmentor
from mmseg.models.data_preprocessor import SegDataPreProcessor

from sam3 import build_sam3_image_model
from sam3.model.sam3_image_processor import Sam3Processor


def get_cls_idx(path: str) -> Tuple[List[str], List[int]]:
    """Parse class names and their corresponding indices from a text file."""
    with open(path, 'r', encoding='utf-8') as f:
        name_sets = f.readlines()
    
    class_names, class_indices = [], []
    for idx, line in enumerate(name_sets):
        names_i = line.strip().split(', ')
        class_names.extend(names_i)
        class_indices.extend([idx] * len(names_i))
        
    return class_names, class_indices


@MODELS.register_module()
class SAM3_OVSS(BaseSegmentor):
    def __init__(self, 
                 classname_path: str, 
                 bpe_path: str = "sam3/assets/bpe_simple_vocab_16e6.txt.gz",
                 checkpoint_path: str = "sam3/assets/sam3.pt",
                 confidence_threshold: float = 0.5,
                 device: str = 'cuda', 
                 **kwargs):
        """
        SAM3 Open-Vocabulary Semantic Segmentation Model.
        """
        data_preprocessor = SegDataPreProcessor(
            mean=[122.771, 116.746, 104.094],
            std=[68.501, 66.632, 70.323],
            bgr_to_rgb=True
        )
        super().__init__(data_preprocessor=data_preprocessor)
        
        self.device = torch.device(device) if isinstance(device, str) else device
        
        # Build SAM3 model and processor
        self.model = build_sam3_image_model(
            bpe_path=bpe_path, 
            checkpoint_path=checkpoint_path, 
            device=device
        )
        self.model.eval()
        self.processor = Sam3Processor(
            self.model, 
            confidence_threshold=confidence_threshold, 
            device=self.device
        )
        
        # Load queries
        self.query_words, self.query_idx = get_cls_idx(classname_path)

    def predict(self, inputs: torch.Tensor, data_samples: Optional[list] = None) -> list:
        if data_samples is not None:
            batch_img_metas = [data_sample.metainfo for data_sample in data_samples]
        else:
            batch_img_metas = [
                dict(
                    ori_shape=inputs.shape[2:],
                    img_shape=inputs.shape[2:],
                    pad_shape=inputs.shape[2:],
                    padding_size=[0, 0, 0, 0]
                )
            ] * inputs.shape[0]
        
        num_cls = max(self.query_idx) + 1
        
        for i, img_meta in enumerate(batch_img_metas): 
            image = Image.open(img_meta['img_path']).convert('RGB')
            ori_shape = img_meta['ori_shape']
            
            seg_logits = torch.zeros((num_cls, *ori_shape), device=self.device)
            
            if data_samples is not None and getattr(data_samples[i], 'reduce_zero_label', False):
                seg_pred = torch.full((1, *ori_shape), 255, dtype=torch.long, device=self.device)
            else:
                seg_pred = torch.zeros((1, *ori_shape), dtype=torch.long, device=self.device)
            
            with torch.no_grad(), torch.autocast(device_type="cuda", dtype=torch.float16):
                inference_state = self.processor.set_image(image)
                
                for query_word, query_idx in zip(self.query_words, self.query_idx):
                    self.processor.reset_all_prompts(inference_state)
                    inference_state = self.processor.set_text_prompt(
                        state=inference_state, 
                        prompt=query_word
                    )
                    
                    inst_len = inference_state['masks_logits'].shape[0]
                    if inst_len > 0:
                        for inst_id in range(inst_len):
                            instance_mask = inference_state['masks'][inst_id].squeeze(dim=0)
                            instance_logits = inference_state['masks_logits'][inst_id].squeeze(dim=0)
                            
                            seg_pred[0][instance_mask] = query_idx
                            seg_logits[query_idx][instance_mask] = instance_logits[instance_mask]
            
            if data_samples is not None:
                data_samples[i].set_data({
                    'seg_logits': PixelData(**{'data': seg_logits}),
                    'pred_sem_seg': PixelData(**{'data': seg_pred})
                })
                
        return data_samples
    
    def _forward(self, inputs, data_samples=None):
        raise NotImplementedError

    def extract_feat(self, inputs):
        raise NotImplementedError

    def encode_decode(self, inputs, batch_img_metas):
        raise NotImplementedError

    def loss(self, inputs, data_samples):
        raise NotImplementedError

```

**Because the provided code is simple, if you find any errors, it is very easy to locate the exact position of the error and find the corresponding solution.** Most issues will likely just point to incorrect file paths or dataset roots in your `base_config.py` or `cfg_***.py`, which are quick to fix.

**Comment by SmallLeishen** [comment]
Thank you for your excellent work. I have successfully reproduced the results of both the baseline and the proposed method in the paper. However, if I want to reproduce the results of Stage 1 and Stage 1 + Stage 2, I am not sure how to set up the configuration files. Could you please provide some guidance?

**Comment by ConceptBank** [comment]
Hi, thank you for your continued interest and for successfully reproducing our main results! 

Since our paper is currently under peer review, the code and instructions we have provided so far are sufficient to reproduce the core experiments. We plan to fully release all remaining experimental code and visualization scripts once the paper's review process is concluded. 

Thank you again for your support and understanding! If you have any further questions, please feel free to reach out.
